src/core/shape_inference/include/topk_shape_inference.hpp

```diff
@@ -0,0 +1,92 @@
+// Copyright (C) 2018-2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include <openvino/core/validation_util.hpp>
+#include <openvino/op/topk.hpp>
+
+#include "utils.hpp"
+
+namespace ov {
+namespace op {
+namespace v1 {
+
+template <typename T>
+void shape_infer(const TopK* op,
+                 const std::vector<T>& input_shapes,
+                 std::vector<T>& output_shapes,
+                 const std::map<size_t, std::shared_ptr<ngraph::runtime::HostTensor>>& constant_data = {}) {
+    using DimType = typename std::iterator_traits<typename T::iterator>::value_type;
+    constexpr bool is_dynamic_shape = std::is_base_of<ov::PartialShape, T>::value;
+
+    NODE_VALIDATION_CHECK(op, (input_shapes.size() == 2 && output_shapes.size() == 2));
+    const auto& input_shape = input_shapes[0];
+    const auto input_rank = input_shape.rank();
+
+    NODE_VALIDATION_CHECK(op,
+                          input_rank.is_dynamic() || input_rank.get_length() > 0,
+                          "Input rank must be greater than 0.");
+
+    const auto& k_shape = input_shapes[1];
+    NODE_VALIDATION_CHECK(op, k_shape.rank().compatible(0), "The 'K' input must be a scalar.");
+
+    auto output_shape = input_shape;
+    if (input_shape.rank().is_static()) {
+        ov::PartialShape k_as_shape;
+        auto normalized_axis = ov::normalize_axis(op, op->get_provided_axis(), input_shape.rank());
```

Comment:
Unfortunately, TopK only exposes `set_axis(const int64_t axis)` as public, we must call `set_axis(const Rank input_rank, const int64_t axis)`  which is protected, and also since it changes member variable of TopK, and out friend template has taken `const TopK *` pointer, so there is noway to modify m_normalized_axis() in static shape cases

Reply:
in summary, op::v1::TopK::validate_and_infer_types() has called set_axis() and m_normalized_axis **is** updated correctly before calling shape_infer(), but in StaticShape inference, there seems to be no way and no reason to update m_normalized_axis in TopK op.